Add JDK and -native JDK builds for other OpenJDK versions

madisongh opened the pull request:
This PR includes some cleanup/simplification of the #57 changes for OpenJDK 21, and adds `-jdk` and `-jdk-native` recipes for the other OpenJDK versions.

This will help folks, like me, that are looking to replace meta-java with this layer and have recipes for their own Java applications that use the OpenJDK 8 JRE from meta-java.

And hopefully the simplification for the jdk and jdk-native support will make the recipes easier to maintain going forward.

madisongh committed:
openjdk-8: add support for jdk and jdk-native builds

Adapted from #57

Signed-off-by: Matt Madison <matt@madison.systems>
Signed-off-by: Sebastian Veit <sebastian.veit@lucimber.com>

madisongh committed:
openjdk-21: update to align with openjdk-8 recipes

Signed-off-by: Matt Madison <matt@madison.systems>

madisongh committed:
openjdk-17: update to align with openjdk-8 recipes

Signed-off-by: Matt Madison <matt@madison.systems>

madisongh committed:
openjdk-11: update to align with openjdk-8 recipes

Signed-off-by: Matt Madison <matt@madison.systems>

sebveit commented:
thanks for your contribution! that's some good addition! i'll verify your changes locally on my pc and merge them into ~~main~~ scarthgap and the other branches after doing so.

madisongh commented:
Thanks, @sebveit ... I have another branch with similar changes for kirkstone, which I was planning on sending a PR for after this goes through, but I can send it sooner if that would be easier.

sebveit commented:
I have the following files and I'm trying to run `bitbake helloworld-java`. I'm probably doing something wrong here, because I get various errors. Can you show me how I can use your provided native-jdk for a recipe?

```
# recipes-java/helloworld-java/helloworld-java_1.0.bb

SUMMARY = "Hello World Java Example"
DESCRIPTION = "A simple Hello World Java program for Yocto"
LICENSE = "MIT"
LIC_FILES_CHKSUM = "file://${COMMON_LICENSE_DIR}/MIT;md5=0835ade698e0bcf8506ecda2f7b4f302"

# javac required at buildtime (define version at local.conf via virtual)
DEPENDS = "openjdk-8-jdk-native"
# JRE required at runtime (select version at local.conf via virtual)
#RDEPENDS:${PN} += "${VIRTUAL-RUNTIME_openjdk-jre}"
RDEPENDS:${PN} += "openjdk-8-jre"

SRC_URI = "file://HelloWorld.java"
S = "${WORKDIR}"

do_compile() {
    ${RECIPE_SYSROOT_NATIVE}/usr/java-native/bin/javac HelloWorld.java
}

do_install() {
    install -d ${D}${bindir}
    echo '#!/bin/sh' > ${D}${bindir}/helloworld-java
    echo 'exec ${bindir}/java -cp ${bindir} HelloWorld' >> ${D}${bindir}/helloworld-java
    chmod +x ${D}${bindir}/helloworld-java

    install -m 0755 HelloWorld.class ${D}${bindir}/
}

```

```
# recipes-java/helloworld-java/files/HelloWorld.java

public class HelloWorld {
    public static void main(String[] args) {
        System.out.println("Hello, World from Java on Yocto!");
    }
}
```

madisongh committed:
helloworld-java: add recipe

for testing cross-compilation of Java code in a recipe.

Signed-off-by: Matt Madison <matt@madison.systems>
Signed-off-by: Sebastian Veit <sebastian.veit@lucimber.com>

madisongh commented:
@sebveit I've added a commit that includes your test recipe and code, with some modifications to make it work. I've built it using stock poky, added it via `CORE_IMAGE_EXTRA_INSTALL` to a core-image-base image, and run it via `runqemu` in qemuarm64.

sebveit commented:
Thank you for the commits regarding the testing of javac with the hello world app. This will come in handy with GitHub Actions for automated testing. As of right now, the compilation isn't working and I don't know why! It throws me unrelated errors. Maybe my setup isn't as clean as I thought. I don't see the forest because of all the trees (German Idiom). I'll continue on this tomorrow.

sebveit commented:
So I'm back on it and it compiles after cleaning the setup (`rm -R cache sstate-cache tmp-glibc`) and removing uneccessary layers. That's good news! It probably failed before because my testing recipe didn't had this line: `PATH =. "${STAGING_LIBDIR_NATIVE}/jvm/openjdk-${JDK_VERSION}-jdk/bin:"`

There might be a better way than adding this line to every Java recipe? I'm looking into it. This might take a while thanks to my potato-pc.

sebveit commented:
BTW my build config looks like this:
```
BB_VERSION           = "2.8.0"
BUILD_SYS            = "x86_64-linux"
NATIVELSBSTRING      = "fedora-40"
TARGET_SYS           = "x86_64-oe-linux"
MACHINE              = "qemux86-64"
DISTRO               = "nodistro"
DISTRO_VERSION       = "nodistro.0"
TUNE_FEATURES        = "m64 core2"
TARGET_FPU           = ""
meta                 = "scarthgap:d1b64f190c1686f081f5ba2c4f2b320048f6a514"
meta-openjdk-temurin = "jdk-for-openjdk-8-scarthgap:b3020fdc8af0c6d489c8f51075797c93859a2f07"
```

madisongh commented:
> There might be a better way than adding this line to every Java recipe? I'm looking into it. 

The meta-java layer has bbclasses for use by recipes that build java libraries and programs, which might provide some inspiration.

sebveit commented:
@madisongh It took me quite some time, but I managed to eliminate the need for an updated PATH environment variable in the helloworld-java example recipe. I’ve split the native and target JDKs into two separate recipes to reduce complexity. Could you please test the changes I’ve introduced? Thanks!

sebveit committed:
ci: Include openjdk-8-jdk and openjdk-8-jdk-native into GitHub workflow.

Signed-off-by: Sebastian Veit <sebastian.veit@lucimber.com>

sebveit commented:
@madisongh just ignore those failing checks. I'm unable to resolve this GitHub-related issue at the moment.

madisongh committed:
openjdk-8: use DEPENDS for dependencies

Rather than explicitly naming libraries and versions
in RDEPENDS, use DEPENDS for recipes with the libraries
and let the build system figure it out for us.

Signed-off-by: Matt Madison <matt@madison.systems>

madisongh committed:
openjdk-8: simplify SRC_URI setting

* Put the two URIs into a single assignment, rather than
  using :append

* Replace the anonymous python function with direct calls
  within the ADAPTED_PV_SRC assignment

Signed-off-by: Matt Madison <matt@madison.systems>

madisongh commented:
@sebveit I implemented an alternative approach to fixing the packaging issue with the multiple copies of the `libjli.so` library. Since the two copies appear to be substantially different, leaving them both in place and listing it in PRIVATE_LIBS is probably safer (and simpler).

I also went ahead and simplified the expression of the dependencies - just using DEPENDS and letting the build system's shared library handling figure out the runtime dependencies during packaging is an approach I've used in other layers that repackage prebuilt binaries.

Finally, as a minor improvement I simplified the SRC_URI setting a bit and eliminated the anonymous Python code from the include file.

I've updated the openjdk-17 and -21 recipes to match the added functionality (what you added plus what I listed above), and have built and tested running `java -version` and `helloworld-java` for each of the supported architectures  for each of openjdk-{8,17,21}. The only issue I ran into was that the JVM won't run on  `qemuriscv64` due to a known compatibility issue; adding `qemuparams="-cpu rv64,sv57=off"` to my `runqemu` invocation bypasses the problem. 

I see you changed the openjdk-8 recipes back to having separate native and target include files and recipes, undoing some of my earlier changes.  I left the -17 and -21 recipes as I had them. If you really prefer going back to the split and the separate -native recipe, I can rebase and drop all my consolidation changes from all of the recipes. Otherwise, I can re-do the consolidation for the -8 recipes.

Thanks

sebveit commented:
> I implemented an alternative approach to fixing the packaging issue with the multiple copies of the libjli.so library. Since the two copies appear to be substantially different, leaving them both in place and listing it in PRIVATE_LIBS is probably safer (and simpler).

I didn't know the trick with `PRIVATE_LIBS` ... this is so much easier!


> I also went ahead and simplified the expression of the dependencies - just using DEPENDS and letting the build system's shared library handling figure out the runtime dependencies during packaging is an approach I've used in other layers that repackage prebuilt binaries.

This is possible? I thought we must stick with `RDEPENDS` since those dependencies are required at runtime by the JVM. Therefore manually defining which packages must be installed in the image along with the JVM.


> Finally, as a minor improvement I simplified the SRC_URI setting a bit and eliminated the anonymous Python code from the include file.

That one is also nice!


> I've updated the openjdk-17 and -21 recipes to match the added functionality (what you added plus what I listed above), and have built and tested running java -version and helloworld-java for each of the supported architectures for each of openjdk-{8,17,21}. The only issue I ran into was that the JVM won't run on qemuriscv64 due to a known compatibility issue; adding qemuparams="-cpu rv64,sv57=off" to my runqemu invocation bypasses the problem.

Thank you for that! Don't worry that much about qemuriscv64, this layer is already in a better shape that the `meta-java` layer and riscv64 is relatively fresh compared to other architectures. Thanks for the hint regarding the qemuparams!


> I see you changed the openjdk-8 recipes back to having separate native and target include files and recipes, undoing some of my earlier changes. I left the -17 and -21 recipes as I had them. If you really prefer going back to the split and the separate -native recipe, I can rebase and drop all my consolidation changes from all of the recipes. Otherwise, I can re-do the consolidation for the -8 recipes.

I get why you combined the target and native recipe for the JDK, but this might become complicated in the future. I think the separation of concerns is useful here. Just look at the include files for the target and native builds — they vary greatly in content. What are your thoughts on this?

I will look into your changes fully on monday! Thank you again for your contribution to this layer!

madisongh commented:
> > I also went ahead and simplified the expression of the dependencies - just using DEPENDS and letting the build system's shared library handling figure out the runtime dependencies during packaging is an approach I've used in other layers that repackage prebuilt binaries.
> 
> This is possible? I thought we must stick with `RDEPENDS` since those dependencies are required at runtime by the JVM. Therefore manually defining which packages must be installed in the image along with the JVM.

Using DEPENDS works for any library dependencies that the code is linked directly against - the `shlibs` processing in the packaging code analyzes all ELF files and generates runtime dependencies for those libraries automatically. You typically only have to use RDEPENDS if there are external libraries that get loaded dynamically at runtime via `dlopen()`.

[...]

> I get why you combined the target and native recipe for the JDK, but this might become complicated in the future. I think the separation of concerns is useful here. Just look at the include files for the target and native builds — they vary greatly in content. What are your thoughts on this?

Not sure what you mean... the differences in the build steps between target and native don't look that different. For target builds, `patchelf --set-interpreter` needs to be run on the programs, which is skipped for native. And for native builds, the symlinks into `${bindir}` get created, which gets skipped for target builds. Those two cases are pretty easily dealt with, though, by using the `:class-native` override in the common include.

OE-Core migrated to using BBCLASSEXTEND and dropping separate -native recipes, for all but a small number of cases, a while back, so it's a well-established pattern these days.

But as I said, I can revert back those changes if that's what you'd prefer.

sebveit commented:
The reason I favor distinct recipes - with shared includes - is that I can spot differences easily. And if the need arise to drop one of those recipes, I simply remove the related file(s). Another important point is bugs. It could happen that by using just one file for target and native recipe, a variable might get set or unset by mistake (caused by me!). If you still think that `:class-native` is a better approach, then please show me how one would combine the following two files:
- recipes-core/openjdk-8/openjdk-8-target.inc
- recipes-core/openjdk-8/openjdk-8-native.inc

madisongh commented:
> It could happen that by using just one file for target and native recipe, a variable might get set or unset by mistake (caused by me!).

Yep, I get it, there's a possibility of introducing errors because you forget that the one file needs to handle both cases. On the other hand, with the multiple-file/include approach, there's the possibility of introducing errors due to forgetting to update both files when a change is needed for both cases.  

The `recipes-core/openjdk-17/openjdk-17.inc` file has the consolidation, and it  would be pretty much the same for openjdk-8. The packaging settings are the same for both cases (with the build system automatically ignoring those not relevant to native builds), and the target/native-specific tweaks to the do_install step are handled there via the override (but the main file-copying step is there just once, since it's the same for both).

But you're the maintainer, and your concern about possible future bugs is the important one here...  I'll work on reverting the -11, -17, and -21 recipes to match the recipe/include layout you have for -8. Will follow up when that's ready.
Thanks

madisongh commented:
@sebveit I've rebased (keeping the native/target split) and retested all versions across all architectures.

sebveit commented:
Perfect! Thank you! I'm merging your PR and close this issue now.
